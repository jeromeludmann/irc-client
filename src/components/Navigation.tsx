import React, { PureComponent, MouseEvent, CSSProperties } from "react";

interface Props {
  servers: {
    [s: string]: {
      windows: { [w: string]: { unread: boolean } };
    };
  };
  active: {
    server: string;
    window: string;
  };
  onWindowButtonClick: (
    { server, window }: { server: string; window: string },
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { active, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(server => (
          <ul key={server} style={{ listStyle: "none", padding: "0" }}>
            {Object.keys(servers[server].windows).map(window => {
              const isActive =
                server === active.server && window === active.window;
              const isUnread = servers[server].windows[window].unread;
              const style: CSSProperties = {
                outline: "none",
                border: "1px solid #ddd",
                width: "100px",
                color: isActive ? "black" : "#bbb",
                backgroundColor: isUnread ? "yellow" : "#fff",
              };

              return (
                <li key={window} style={{ marginBottom: "5px" }}>
                  <button
                    onClick={this.handleClick}
                    data-server={server}
                    data-window={window}
                    style={style}
                  >
                    {window}
                  </button>
                </li>
              );
            })}
          </ul>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const server = e.currentTarget.getAttribute("data-server");
    const window = e.currentTarget.getAttribute("data-window");

    if (server && window) {
      this.props.onWindowButtonClick({ server, window });
    }
  };
}
