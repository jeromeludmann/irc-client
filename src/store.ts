import { createStore, applyMiddleware } from 'redux'
import { messageParser } from '@app/middlewares/messageParser'
import { commandHandler } from '@app/middlewares/commandHandler'
import { socketHandler } from '@app/middlewares/socketHandler'
import { autoRouter } from '@app/middlewares/autoRouter'
import { lag } from '@app/middlewares/lag'
import { pingPong } from '@app/middlewares/pingPong'
import { register } from '@app/middlewares/register'
import { logger } from '@app/middlewares/logger'
import { windowHandler } from '@app/middlewares/windowHandler'
import { reduceRoot, rootInitialState } from '@app/state/root/reducer'

export const store = createStore(
  reduceRoot,
  rootInitialState,
  applyMiddleware(
    messageParser, // keep first
    autoRouter,
    lag,
    pingPong,
    register,
    commandHandler,
    windowHandler,
    socketHandler, // keep just before logger
    logger, // keep last
  ),
)
