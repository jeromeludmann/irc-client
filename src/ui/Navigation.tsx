import React, { PureComponent, MouseEvent } from "react";

interface Props {
  servers: {
    [server: string]: {
      channels: {
        [channel: string]: {
          unread: boolean;
        };
      };
    };
  };
  active: {
    server: string;
    channel: string;
  };
  onChannelClick: (server: string, channel: string) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { active, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(server => (
          <ul key={server}>
            {Object.keys(servers[server].channels).map(channel => {
              const isActive =
                server === active.server && channel === active.channel;
              // const isUnread = servers[server].channels[channel].unread;

              return (
                <li key={`${server}/${channel}`}>
                  <button
                    onClick={this.handleClick}
                    data-server={server}
                    data-channel={channel}
                  >
                    {channel === "status" ? server : channel}{" "}
                    {isActive && <span>&lt;</span>}
                  </button>
                </li>
              );
            })}
          </ul>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const server = e.currentTarget.getAttribute("data-server");
    const channel = e.currentTarget.getAttribute("data-channel"); // explicitly set to undefined if null

    if (server && channel) {
      this.props.onChannelClick(server, channel);
    }
  };
}
