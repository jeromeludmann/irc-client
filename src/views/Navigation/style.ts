import styled, { css } from 'styled-components'

export const List = styled.ul`
  margin-bottom: 10px;
  padding: 0;
  list-style: none;
`

export const ListItem = styled.li`
  margin: 5px 0;
  list-style: none;
`

interface ButtonProps {
  isActive: boolean
  hasActivity: boolean
}

export const Button = styled.button`
  outline: none;
  border: 1px solid #ddd;
  width: 100%;
  height: 30px;
  cursor: pointer;
  color: #bbb;
  background-color: white;
  font-weight: normal;
  &:hover {
    color: red;
  }

  ${(props: ButtonProps) =>
    props.isActive &&
    css`
      color: #333;
      font-weight: bold;
    `};

  ${(props: ButtonProps) =>
    props.hasActivity &&
    css`
      background-color: yellow;
    `};
`
