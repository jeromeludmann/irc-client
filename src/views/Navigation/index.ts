import { connect, MapStateToProps } from 'react-redux'
import { RouteState } from '@app/state/route/reducer'
import { Route } from '@app/utils/Route'
import { RootState } from '@app/state/root/reducer'
import { switchWindow } from '@app/actions/ui'
import Navigation from '@app/views/navigation/Navigation'
import { selectServers } from '@app/state/root/selectors'
import { selectRoute } from '@app/state/route/selectors'

interface StateProps {
  servers: {
    [serverKey: string]: {
      name: string
      buffers: { [bufferKey: string]: { activity: boolean } }
    }
  }
  window: RouteState
}

interface DispatchProps {
  onWindowButtonClick: (route: Route) => void
}

const mapStateToProps: MapStateToProps<StateProps, {}, RootState> = state => ({
  // TODO provide a better selector to Navigation
  servers: selectServers(state),
  window: selectRoute(state),
})

const mapDispatchToProps: DispatchProps = {
  onWindowButtonClick: switchWindow,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Navigation)
