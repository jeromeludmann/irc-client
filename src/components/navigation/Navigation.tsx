import React, { PureComponent, MouseEvent } from "react";
import { STATUS, RAW } from "@app/Route";
import { List, ListItem, Button } from "@app/components/navigation/style";

interface Props {
  servers: {
    [serverKey: string]: {
      name: string;
      channels: { [channelKey: string]: { activity: boolean } };
    };
  };
  window: {
    serverKey: string;
    channelKey: string;
  };
  onChannelButtonClick: (
    { serverKey, channelKey }: { serverKey: string; channelKey: string }
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { window, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(serverKey => (
          <List key={serverKey}>
            {Object.keys(servers[serverKey].channels).map(channelKey => (
              <ListItem key={channelKey}>
                <Button
                  onClick={this.handleClick}
                  data-server-key={serverKey}
                  data-channel-key={channelKey}
                  isActive={
                    serverKey === window.serverKey &&
                    channelKey === window.channelKey
                  }
                  hasActivity={servers[serverKey].channels[channelKey].activity}
                >
                  {channelKey === STATUS
                    ? servers[serverKey].name
                    : channelKey === RAW
                      ? serverKey
                      : channelKey}
                </Button>
              </ListItem>
            ))}
          </List>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const serverKey = e.currentTarget.getAttribute("data-server-key") as string;
    const channelKey = e.currentTarget.getAttribute(
      "data-channel-key"
    ) as string;

    this.props.onChannelButtonClick({ serverKey, channelKey });
  };
}
