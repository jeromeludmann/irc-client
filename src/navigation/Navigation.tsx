import React, { Component, MouseEvent } from "react";
import { connect } from "react-redux";
import { RootState } from "@app/state";
import { getServers } from "@app/state/servers";
import { setActiveWindow } from "@app/navigation/actions";

interface StateProps {
  current: {
    server: string;
    channel: string;
  };
  servers: {
    [server: string]: {
      channels: {
        [channel: string]: {
          unread: boolean;
        };
      };
    };
  };
}

interface DispatchProps {
  onChannelClick: (server: string, channel: string) => void;
}

class Navigation extends Component<StateProps & DispatchProps> {
  public render() {
    const { current, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(server => (
          <ul
            key={server}
            style={{
              listStyle: "none",
              padding: 0,
              marginTop: "6px",
            }}
          >
            {Object.keys(servers[server].channels).map(channel => {
              const active =
                server === current.server && channel === current.channel;
              const unread = servers[server].channels[channel].unread;

              return (
                <li key={`${server}/${channel}`} style={{ marginTop: "3px" }}>
                  <button
                    onClick={this.handleClick}
                    data-server={server}
                    data-channel={channel}
                    style={{
                      backgroundColor: active ? "#ccc" : "#333",
                      color: active ? "#111" : "#ccc",
                      fontWeight: unread ? "bold" : "normal",
                      outline: "none",
                      width: "100%",
                      height: "24px",
                      border: 0,
                      fontSize: channel ? "12px" : "13px",
                    }}
                  >
                    {channel === "status" ? server : channel}
                  </button>
                </li>
              );
            })}
          </ul>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const server = e.currentTarget.getAttribute("data-server");
    const channel = e.currentTarget.getAttribute("data-channel"); // explicitly set to undefined if null

    if (server && channel) {
      this.props.onChannelClick(server, channel);
    }
  };
}

const mapStateToProps = (state: RootState): StateProps => ({
  current: state.current,
  servers: getServers(state),
});

const mapDispatchToProps = {
  onChannelClick: setActiveWindow,
};

export default connect<StateProps, DispatchProps>(
  mapStateToProps,
  mapDispatchToProps,
)(Navigation);
