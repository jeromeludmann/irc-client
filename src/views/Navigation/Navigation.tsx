import React, { PureComponent, MouseEvent } from 'react'
import { isStatus, isRaw } from '@app/utils/Route'
import { List, ListItem, Button } from '@app/views/navigation/style'

interface Props {
  servers: {
    [serverKey: string]: {
      name: string
      buffers: { [buffer: string]: { activity: boolean } }
    }
  }
  window: {
    serverKey: string
    bufferKey: string
  }
  onWindowButtonClick: (
    { serverKey, bufferKey }: { serverKey: string; bufferKey: string },
  ) => void
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { window, servers } = this.props

    return (
      <>
        {Object.keys(servers).map(serverKey => (
          <List key={serverKey}>
            {Object.keys(servers[serverKey].buffers).map(bufferKey => (
              <ListItem key={bufferKey}>
                <Button
                  onClick={this.handleClick}
                  data-server-key={serverKey}
                  data-buffer-key={bufferKey}
                  isActive={
                    serverKey === window.serverKey &&
                    bufferKey === window.bufferKey
                  }
                  hasActivity={servers[serverKey].buffers[bufferKey].activity}
                >
                  {isStatus(bufferKey)
                    ? servers[serverKey].name
                    : isRaw(bufferKey)
                      ? serverKey
                      : bufferKey}
                </Button>
              </ListItem>
            ))}
          </List>
        ))}
      </>
    )
  }

  private handleClick = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    this.props.onWindowButtonClick({
      serverKey: currentTarget.getAttribute('data-server-key')!,
      bufferKey: currentTarget.getAttribute('data-buffer-key')!,
    })
  }
}
