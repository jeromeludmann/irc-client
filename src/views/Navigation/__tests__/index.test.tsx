import React from 'react'
import { shallow } from 'enzyme'
import Navigation from '@app/views/Navigation'
import { store } from '@app/store'

describe('Navigation container', () => {
  it('should render correctly', () => {
    expect(shallow(<Navigation />, { context: { store } })).toMatchSnapshot()
  })
})
