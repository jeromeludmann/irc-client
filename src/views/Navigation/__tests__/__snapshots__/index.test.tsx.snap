// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation container should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Navigation) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "onWindowButtonClick": [Function],
      "servers": Object {
        "serverKey": Object {
          "buffers": Object {
            "@raw": Object {
              "activity": false,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 0,
                  "values": Array [],
                },
                "value": "",
              },
              "messages": Array [],
            },
            "@status": Object {
              "activity": false,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 0,
                  "values": Array [],
                },
                "value": "",
              },
              "messages": Array [],
            },
          },
          "lag": 0,
          "modes": Object {
            "available": Object {
              "channel": Array [],
              "user": Array [],
            },
            "user": Array [],
          },
          "name": "<unknown>",
          "user": Object {
            "nick": "default_nick",
            "real": "default_name",
            "user": "default_user",
          },
        },
      },
      "window": Object {
        "bufferKey": "@status",
        "serverKey": "serverKey",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "onWindowButtonClick": [Function],
        "servers": Object {
          "serverKey": Object {
            "buffers": Object {
              "@raw": Object {
                "activity": false,
                "input": Object {
                  "dirtyValue": "",
                  "history": Object {
                    "index": 0,
                    "values": Array [],
                  },
                  "value": "",
                },
                "messages": Array [],
              },
              "@status": Object {
                "activity": false,
                "input": Object {
                  "dirtyValue": "",
                  "history": Object {
                    "index": 0,
                    "values": Array [],
                  },
                  "value": "",
                },
                "messages": Array [],
              },
            },
            "lag": 0,
            "modes": Object {
              "available": Object {
                "channel": Array [],
                "user": Array [],
              },
              "user": Array [],
            },
            "name": "<unknown>",
            "user": Object {
              "nick": "default_nick",
              "real": "default_name",
              "user": "default_user",
            },
          },
        },
        "window": Object {
          "bufferKey": "@status",
          "serverKey": "serverKey",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
    },
    "hydrateIn": undefined,
  },
}
`;
